utils.go

```diff
@@ -0,0 +1,168 @@
+package loop
+
+import (
+	"context"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+	"github.com/lightningnetwork/lnd/zpay32"
+)
+
+var (
+	// DefaultMaxHopHints is set to 20 as that is the default set in LND
+	DefaultMaxHopHints = 20
+)
+
+// SelectHopHints is a direct port of the SelectHopHints found in lnd. It was
+// reimplemented because the current implementation in LND relies on internals
+// not externalized through the API. Hopefully in the future SelectHopHints
+// will be refactored to allow for custom data sources. It iterates through all
+// the active and public channels available and returns eligible channels.
+// Eligibility requirements are simple: does the channel have enough liquidity
+// to fulfill the request and is the node whitelisted (if specified)
+func SelectHopHints(ctx context.Context, lnd *lndclient.LndServices,
+	amtMSat btcutil.Amount, numMaxHophints int,
+	includeNodes map[route.Vertex]struct{}) ([][]zpay32.HopHint, error) {
+
+	// Fetch all active and public channels.
+	openChannels, err := lnd.Client.ListChannels(ctx, false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// We'll add our hop hints in two passes, first we'll add all channels
+	// that are eligible to be hop hints, and also have a local balance
+	// above the payment amount.
+	var totalHintBandwidth btcutil.Amount
+	hopHintChans := make(map[int]*lndclient.ChannelEdge)
+	hopHints := make([][]zpay32.HopHint, 0, numMaxHophints)
+
+	for i, channel := range openChannels {
+		// In this first pass, we'll ignore all channels in
+		// isolation can't satisfy this payment.
+
+		// Retrieve extra info for each channel not available in
+		// listChannels
+		chanInfo, err := lnd.Client.GetChanInfo(ctx, channel.ChannelID)
+		if err != nil {
+			return nil, err
+		}
+		// Cache the GetChanInfo result since it might be useful
+		hopHintChans[i] = chanInfo
+
+		//Skip if channel can't forward payment
+		if channel.RemoteBalance < amtMSat {
+			log.Infof(
+				"Skipping ChannelID: %v for hints as remote balance (%v sats) is insufficient"+
+					"appears to be private",
+				channel.ChannelID, channel.RemoteBalance,
+			)
+			continue
+		}
+		// If includeNodes is set, we'll only add channels with peers in
+		// include_node. This is done to respect the last_hop parameter
+		if len(includeNodes) > 0 {
+			if _, ok := includeNodes[channel.PubKeyBytes]; !ok {
+				continue
+			}
+		}
+
+		// We want to prevent leaking private nodes, which we define as
+		// nodes with only private channels
+		//
+		// GetNodeInfo will never return private channels, even if
+		// they're somehow known to us. If there are any channels
+		// returned, we can consider the node to be public.
+		nodeInfo, err := lnd.Client.GetNodeInfo(
```

Comment:
nitty: This may potentially result in a lot of `GetNodeInfo` calls. How about just filtering the `openChannels` so it doesn't include channels to private nodes?

Reply:
Also just a thought: if a node only has private channels, we won't be able to loopin there. I'd say we also need to support that? wdyt @carlaKC ?

Reply:
The issue here is that we're trying to mimic lnd's "private node" behaviour, where if we're connected to a node by a private channel which has no other public channels, then we're nice and we don't dox them by adding a hop hint. In any case, adding that hop hint wouldn't be useful, because there are no channels to route to the private node and then to us. 

Agree on lots of `GetNodeInfo` calls, since we're probably checking our same set of peers over and over. Can probably add come kind of cache which lives for 10 mins (block time)? Not a disaster if it's a bit stale. Can probably be in a followup though, @bhandras?

Reply:
> The issue here is that we're trying to mimic lnd's "private node" behaviour, where if we're connected to a node by a private channel which has no other public channels, then we're nice and we don't dox them by adding a hop hint. In any case, adding that hop hint wouldn't be useful, because there are no channels to route to the private node and then to us.

Yeah, indeed!

> Agree on lots of GetNodeInfo calls, since we're probably checking our same set of peers over and over. Can probably add come kind of cache which lives for 10 mins (block time)? Not a disaster if it's a bit stale. 

Now that I thought this through again, I was wrong here. I meant to just use the `openChannels` but indeed we need to know if the peer is fully private so we need to query. I think the caching is not that important in this case since we'd only spare a few calls if we have multiple channels to fully private nodes.